Feature/raja chai umpire update

adrienbernede opened the pull request:
This PR supersedes #12959

### The goals:
- Add CHAI spack package.
- Reference this package as an optional dependency in RAJA.
- Update the previous PR and address the review comments.

### Notes:
- Umpire is only marginally impacted.
- I removed mentions of TBB and OpenMP backend. I haven’t tested this so I preferred not adding it.
- CHAI package was not thoroughly tested, but we need this package to introduce better CI on our side.

@rblake-llnl @davidbeckingsale @rhornung67

Closes #12959

rblake-llnl committed:
Changing raja, chai, and umpire packages so all will compile with each other.

rblake-llnl committed:
Need a CUDA version of CHAI when compiling with raja+cuda+chai

rblake-llnl committed:
Updating checks for commit.

rblake-llnl committed:
Adding comments explaining why chai+umpire tests were disabled

adrienbernede committed:
Reactivating tests for CHAI and Umpire

adrienbernede committed:
reordering versions

adrienbernede committed:
Unified handling of Cuda Arch

adrienbernede committed:
Adding latest versions

adrienbernede committed:
Unused/Untested: removed

adrienbernede committed:
Aesthetic and test mode in Chai

adrienbernede committed:
Unified handling of Cuda Arch

rblake-llnl commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -61,4 +71,8 @@ def cmake_args(self):
 
+        if '+chai' in spec:
+            options.extend([
+                '-DENABLE_CHAI=On'])
```
Needs to be ON

adrienbernede replied:
Done

rblake-llnl commented on var/spack/repos/builtin/packages/chai/package.py:
```diff
@@ -0,0 +1,56 @@
+    version('1.1.0', tag='v1.1.0', submodules='True')
+    version('1.0', tag='v1.0', submodules='True')
+
+    depends_on('cmake@3.8:', type='build')
```
Do we need a variant for CUDA support?  How will the user know that they need to select both +cuda and +cuda_arch ?

adrienbernede replied:
Cuda variant is added by heritage of CudaPackage class.

rblake-llnl replied:
Cool!  I didn't know that existed!  Does that pull in cmake package options as well?

adrienbernede replied:
This is done by double inheritance:
https://github.com/spack/spack/blob/ce6126d8ce4bc40f688932c0df9385fb3c57582c/var/spack/repos/builtin/packages/chai/package.py#L9

rblake-llnl commented on var/spack/repos/builtin/packages/raja/package.py:
```diff
@@ -30,9 +30,12 @@ class Raja(CMakePackage, CudaPackage):
     depends_on('cmake@3.8:', type='build')
     depends_on('cmake@3.9:', when='+cuda', type='build')
+    depends_on('chai', when='+chai')
+    depends_on('chai+cuda', when='+chai+cuda')
```
This is my spack ignorance talking, but how will chai+cuda properly carry the +cuda_arch flag that needs to be specified?

adrienbernede replied:
This is based on https://spack.readthedocs.io/en/latest/build_systems/cudapackage.html#usage

Apparently cuda_arch is automatically detected by CudaPackage.

adamjstewart replied:
`cuda_arch` is not (yet) automatically detected, it needs to be set by users. I recommend setting it once in your `packages.yaml` so it applies to all packages.

adrienbernede committed:
Using 'ON' consistently, instead of 'On'

adamjstewart commented on var/spack/repos/builtin/packages/chai/package.py:
```diff
@@ -0,0 +1,56 @@
+    git      = "https://github.com/LLNL/CHAI.git"
+
+    version('develop', branch='develop', submodules='True')
+    version('main', branch='main', submodules='True')
```
```suggestion
    version('master', branch='main', submodules='True')
```
This will allow it to sort correctly, since main is presumably newer than all stable releases.

adamjstewart commented on var/spack/repos/builtin/packages/chai/package.py:
```diff
@@ -0,0 +1,56 @@
+            options.append('-DENABLE_CUDA=OFF')
+
+        options.append('-Dumpire_DIR:PATH='
+                       + spec['umpire'].prefix + "/share/umpire/cmake")
```
```suggestion
                       + spec['umpire'].prefix.share.umpire.cmake)
```

adrienbernede committed:
Apply suggestions from code review

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

adrienbernede committed:
Fix, suggestion and patch:

Chai depends on RAJA, not the other way.
Apply suggested master-main version mapping.
Add Umpire version 3.0.0 and patch.
